url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/Tffst.java

```diff
@@ -0,0 +1,1712 @@
+                      }
+                    }
+                  }
+                }
+              } else {
+                for (Transition tNext : t.getTo().getTransitions()) {
+                  if (!(tNext.labelIn.isEpsilon() && tNext.getTo().equals(tNext.getFrom()))) {
+                    TfString lbout = tNext.labelOut.clone();
+                    for (int i = 0; i < t.labelOut.size(); i++)
+                      lbout.add(i, t.labelOut.get(i));
+                    Transition tNextClone = new Transition(null, tNext.labelIn.clone(), lbout, tNext.getTo());
+                    toAdd.add(tNextClone);
+                    removeState = false;
+                  }
+                }
+              }
+            }
+          }
+        }
+        state.removeAllTransitions(toRemove);
+        state.addAllTransitions(toAdd);
+        if (removeState) workingStates.remove(state);
+      }
+    } catch (CloneNotSupportedException e) {
+      e.printStackTrace();
+    }
+    
+    removeDeadTransitions();
+  }
+
+  
+  
+  /**
+   * returns a Tffsr representing this tffst. it expands subEpochs in simpe
+   * pairts of TFs adding epsilons
+   * 
+   * @return
+   */
+  public Tffsr firstProjection() {
+   
+    Tffsr outTffsr = new Tffsr();
+    Map<State, uy.edu.fing.mina.fsa.tffsr.State> stateMap = new HashMap<State, uy.edu.fing.mina.fsa.tffsr.State>();
+
+    for (State fstState : getStates()) {
+      uy.edu.fing.mina.fsa.tffsr.State fsrState = new uy.edu.fing.mina.fsa.tffsr.State();
+      fsrState.setAccept(fstState.isAccept());
+      if (initial.equals(fstState)) 
+        outTffsr.initial = fsrState;
+      stateMap.put(fstState, fsrState);
+    }
+    
+    for (State fstState : stateMap.keySet()) {
+      uy.edu.fing.mina.fsa.tffsr.State fsrState = stateMap.get(fstState);
+      for (Transition fstTransition :fstState.getTransitions()) {
+        // can do this because  simpleTffst = toSimpleTransitions() 
+        fsrState.addOutTran(
+        		new uy.edu.fing.mina.fsa.tffsr.Transition(
+        			new TfString(fstTransition.labelIn.isEpsilon() ? SimpleTf.Epsilon() : fstTransition.labelIn.get(0)), 
+        				stateMap.get(fstTransition.getTo())));
+      }
+    }
+    return outTffsr;
+    
+  }
+
+  /**
+   * Returns the set of reachable accept states.
+   * 
+   * @return set of {@link State}objects
+   */
+  public Set<State> getAcceptStates() {
+    HashSet<State> accepts = new HashSet<State>();
+    HashSet<State> visited = new HashSet<State>();
+    LinkedList<State> worklist = new LinkedList<State>();
+    worklist.add(initial);
+    visited.add(initial);
+    while (worklist.size() > 0) {
+      State s = worklist.removeFirst();
+      if (s.accept) accepts.add(s);
+      Iterator<Transition> i = s.getTransitionsIterator();
+      while (i.hasNext()) {
+        Transition t = i.next();
+        if (!visited.contains(t.getTo())) {
+          visited.add(t.getTo());
+          worklist.add(t.getTo());
+        }
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 727:733, 812:818.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.